File: backend/models/message.go
package models

import "go.mongodb.org/mongo-driver/bson/primitive"

type Message struct {
    ID                primitive.ObjectID   `bson:"_id,omitempty" json:"id"`
    ChatID            primitive.ObjectID   `bson:"chatId" json:"chatId"`
    SenderID          primitive.ObjectID   `bson:"senderId" json:"senderId"`
    Content           string               `bson:"content" json:"content"`
    Type              string               `bson:"type" json:"type"` // text, image, voice
    IsRead            bool                 `bson:"isRead" json:"isRead"`
    Reactions         map[string]string    `bson:"reactions,omitempty" json:"reactions,omitempty"`
    CreatedAt         int64                `bson:"createdAt" json:"createdAt"`
    ReplyToID         string               `bson:"replyToId,omitempty" json:"replyToId,omitempty"`
    ReplyToContent    string               `bson:"replyToContent,omitempty" json:"replyToContent,omitempty"`
    ReplyToSenderName string               `bson:"replyToSenderName,omitempty" json:"replyToSenderName,omitempty"`
}

File: mobile_app/lib/models/message.dart
class Message {
  final String id;
  final String senderId;
  final String content;
  final String type;
  final DateTime createdAt;
  bool isRead;
  final Map<String, String>? reactions;
  final String? replyToId;
  final String? replyToContent;
  final String? replyToSenderName;

  Message({
    required this.id,
    required this.senderId,
    required this.content,
    this.type = 'text',
    required this.createdAt,
    this.isRead = false,
    this.reactions,
    this.replyToId,
    this.replyToContent,
    this.replyToSenderName,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    // Handle createdAt as int (Unix timestamp) or String
    DateTime parseCreatedAt(dynamic value) {
      if (value is int) {
        return DateTime.fromMillisecondsSinceEpoch(value * 1000);
      } else if (value is String) {
        return DateTime.tryParse(value) ?? DateTime.now();
      }
      return DateTime.now();
    }

    // Parse reactions map
    Map<String, String>? parseReactions(dynamic value) {
      if (value is Map<String, dynamic>) {
        return value.map((key, val) => MapEntry(key, val.toString()));
      }
      return null;
    }

    return Message(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      senderId: json['senderId']?.toString() ?? json['sender']?['_id']?.toString() ?? '',
      content: json['content']?.toString() ?? '',
      type: json['type']?.toString() ?? 'text',
      createdAt: parseCreatedAt(json['createdAt']),
      isRead: json['isRead'] == true,
      reactions: parseReactions(json['reactions']),
      replyToId: json['replyToId']?.toString(),
      replyToContent: json['replyToContent']?.toString(),
      replyToSenderName: json['replyToSenderName']?.toString(),
    );
  }
}
